fix: increase consistency of rpcauth parsing

tdb3 opened the pull request:
The current `rpcauth` parsing behavior is inconsistent and unintuitive (see https://github.com/bitcoin/bitcoin/pull/29141#issuecomment-1972085251 and additional details below).
The current behavior inconsistently treats empty `rpcauth` as an error (or not) depending on the location within CLI/bitcoin.conf and the location of adjacent valid `rpcauth` params.

Empty `rpcauth` is now consistently treated as an error and prevents bitcoind from starting.
Continuation of the upforgrabs PR #29141.


### Additional details:
Current `rpcauth` behavior is nonsensical:

 - If an empty `rpcauth` argument was specified as the last command line argument, it would cause all other `rpcauth` arguments to be ignored.
 - If an empty `rpcauth` argument was specified on the command line followed by any nonempty `rpcauth` argument, it would cause an error.
 - If an empty `rpcauth=` line was specified after non-empty rpcauth line in the config file it would cause an error.
 - If an empty `rpcauth=` line in a config file was first it would cause other rpcauth entries in the config file to be ignored, unless there were `-rpcauth` command line arguments and the last one was nonempty, in which case it would cause an error.

New behavior is simple:
 - If an empty rpcauth config line or command line argument is used it will cause an error

tdb3 commented:
## Behavior (master (bd5d1688b4311e21c0e0ff89a3ae02ef7d0543b8) vs PR):

### CLI:
1) one valid rpcauth
```
src/bitcoind -regtest -rpcauth=testuser1:798e64e1173b5fa3a5946fbd42ced60e\$8472c4ab034459f7c83f59d94f26c820179455946023fa77a38bc51c785573c4
```
current: bitcoind starts, rpcauth usable
new: no change

2) one empty rpcauth
```
src/bitcoind -regtest -rpcauth
src/bitcoind -regtest -rpcauth=
src/bitcoind -regtest -rpcauth=""
```
current: bitcoind starts
new: bitcoind fails, `Invalid (empty) -rpcauth argument`

3) valid rpcauth, then empty rpcauth
```
src/bitcoind -regtest -rpcauth=testuser1:798e64e1173b5fa3a5946fbd42ced60e\$8472c4ab034459f7c83f59d94f26c820179455946023fa77a38bc51c785573c4 -rpcauth
src/bitcoind -regtest -rpcauth=testuser1:798e64e1173b5fa3a5946fbd42ced60e\$8472c4ab034459f7c83f59d94f26c820179455946023fa77a38bc51c785573c4 -rpcauth=
src/bitcoind -regtest -rpcauth=testuser1:798e64e1173b5fa3a5946fbd42ced60e\$8472c4ab034459f7c83f59d94f26c820179455946023fa77a38bc51c785573c4 -rpcauth=""
```
current: bitcoind starts, valid rpcauth ignored
new: bitcoind fails, `Invalid (empty) -rpcauth argument`

4) empty rpcauth, then valid rpcauth
```
src/bitcoind -regtest -rpcauth -rpcauth=testuser1:798e64e1173b5fa3a5946fbd42ced60e\$8472c4ab034459f7c83f59d94f26c820179455946023fa77a38bc51c785573c4
src/bitcoind -regtest -rpcauth= -rpcauth=testuser1:798e64e1173b5fa3a5946fbd42ced60e\$8472c4ab034459f7c83f59d94f26c820179455946023fa77a38bc51c785573c4
src/bitcoind -regtest -rpcauth="" -rpcauth=testuser1:798e64e1173b5fa3a5946fbd42ced60e\$8472c4ab034459f7c83f59d94f26c820179455946023fa77a38bc51c785573c4
```
current: bitcoind fails, `Invalid -rpcauth argument`
new: bitcoind fails, `Invalid (empty) -rpcauth argument`

5) valid rpcauth, empty rpcauth, then valid rpcauth
```
src/bitcoind -regtest -rpcauth=testuser2:05b1fe59e25eec2826be285ae71cf943\$e9035aad77d8c85044e2bca366c38bc9105d7be3447591937add8faaadfb454c -rpcauth -rpcauth=testuser1:798e64e1173b5fa3a5946fbd42ced60e\$8472c4ab034459f7c83f59d94f26c820179455946023fa77a38bc51c785573c4
src/bitcoind -regtest -rpcauth=testuser2:05b1fe59e25eec2826be285ae71cf943\$e9035aad77d8c85044e2bca366c38bc9105d7be3447591937add8faaadfb454c -rpcauth= -rpcauth=testuser1:798e64e1173b5fa3a5946fbd42ced60e\$8472c4ab034459f7c83f59d94f26c820179455946023fa77a38bc51c785573c4
src/bitcoind -regtest -rpcauth=testuser2:05b1fe59e25eec2826be285ae71cf943\$e9035aad77d8c85044e2bca366c38bc9105d7be3447591937add8faaadfb454c -rpcauth="" -rpcauth=testuser1:798e64e1173b5fa3a5946fbd42ced60e\$8472c4ab034459f7c83f59d94f26c820179455946023fa77a38bc51c785573c4
```
current: bitcoind fails, `Invalid -rpcauth argument`
new: bitcoind fails, `Invalid (empty) -rpcauth argument`

### bitcoin.conf:
1) one valid rpcauth
```
rpcauth=testuser1:798e64e1173b5fa3a5946fbd42ced60e$8472c4ab034459f7c83f59d94f26c820179455946023fa77a38bc51c785573c4
```
current: bitcoind starts, rpcauth usable
new: no change

2) one empty rpcauth
```
rpcauth
```
current: bitcoind fails, `parse error on line X: rpcauth`
new: no change

```
rpcauth=
```
current: bitcoind starts
new: bitcoind fails, `Invalid (empty) -rpcauth argument`

```
rpcauth=""
```
current: bitcoind fails, `Invalid -rpcauth argument`
new: no change

3) valid rpcauth, then empty rpcauth
```
rpcauth=testuser1:798e64e1173b5fa3a5946fbd42ced60e$8472c4ab034459f7c83f59d94f26c820179455946023fa77a38bc51c785573c4
rpcauth
```
current: bitcoind fails, `parse error on line X: rpcauth`
new: no change

```
rpcauth=testuser1:798e64e1173b5fa3a5946fbd42ced60e$8472c4ab034459f7c83f59d94f26c820179455946023fa77a38bc51c785573c4
rpcauth=
```
current: bitcoind fails, `Invalid -rpcauth argument`
new: bitcoind fails, `Invalid (empty) -rpcauth argument`

```
rpcauth=testuser1:798e64e1173b5fa3a5946fbd42ced60e$8472c4ab034459f7c83f59d94f26c820179455946023fa77a38bc51c785573c4
rpcauth=""
```
current: bitcoind fails, `Invalid -rpcauth argument`
new: no change

4) empty rpcauth, then valid rpcauth
```
rpcauth
rpcauth=testuser1:798e64e1173b5fa3a5946fbd42ced60e$8472c4ab034459f7c83f59d94f26c820179455946023fa77a38bc51c785573c4
```
current: bitcoind fails, `parse error on line X: rpcauth`
new: no change

```
rpcauth=
rpcauth=testuser1:798e64e1173b5fa3a5946fbd42ced60e$8472c4ab034459f7c83f59d94f26c820179455946023fa77a38bc51c785573c4
```
current: bitcoind starts, valid rpcauth ignored
new: bitcoind fails, `Invalid (empty) -rpcauth argument`

```
rpcauth=""
rpcauth=testuser1:798e64e1173b5fa3a5946fbd42ced60e$8472c4ab034459f7c83f59d94f26c820179455946023fa77a38bc51c785573c4
```
current: bitcoind fails, `Invalid -rpcauth argument`
new: no change

5) valid rpcauth, empty rpcauth, then valid rpcauth
```
rpcauth=testuser2:05b1fe59e25eec2826be285ae71cf943$e9035aad77d8c85044e2bca366c38bc9105d7be3447591937add8faaadfb454c
rpcauth
rpcauth=testuser1:798e64e1173b5fa3a5946fbd42ced60e$8472c4ab034459f7c83f59d94f26c820179455946023fa77a38bc51c785573c4
```
current: bitcoind fails, `parse error on line X: rpcauth`
new: no change

```
rpcauth=testuser2:05b1fe59e25eec2826be285ae71cf943$e9035aad77d8c85044e2bca366c38bc9105d7be3447591937add8faaadfb454c
rpcauth=
rpcauth=testuser1:798e64e1173b5fa3a5946fbd42ced60e$8472c4ab034459f7c83f59d94f26c820179455946023fa77a38bc51c785573c4
```
current: bitcoind fails, `Invalid -rpcauth argument`
new: bitcoind fails, `Invalid (empty) -rpcauth argument`

```
rpcauth=testuser2:05b1fe59e25eec2826be285ae71cf943$e9035aad77d8c85044e2bca366c38bc9105d7be3447591937add8faaadfb454c
rpcauth=""
rpcauth=testuser1:798e64e1173b5fa3a5946fbd42ced60e$8472c4ab034459f7c83f59d94f26c820179455946023fa77a38bc51c785573c4
```
current: bitcoind fails, `Invalid -rpcauth argument`
new: no change

tdb3 commented:
`test each commit` CI fails on f6d0df30a460db75a944ccea38dfb9fd0827042b to help show current behavior.  I can swap the order of the commits to make it pass if need be.

https://github.com/bitcoin/bitcoin/pull/29141#issuecomment-1972085251
> it would be a lot more obvious what current behavior is and how the bugfix changes it.

luke-jr commented:
Does this consider the scenario where bitcoin.conf has valid rpcauth params, but the user specifies -norpcauth or such on the command line to disable them?

tdb3 commented:
> Does this consider the scenario where bitcoin.conf has valid rpcauth params, but the user specifies -norpcauth or such on the command line to disable them?

Thanks for taking a look.
Yes.
`conf_setup()` creates `rpcauth` statements in node 0's bitcoin.conf (including for user `rt`).
After restarting the node with `-norpcauth`, the `assert_equal` with `rt`  tests that the associated `rpcauth` statement in the conf is disabled.

https://github.com/bitcoin/bitcoin/blob/1ef15ea7120fe3f048531ae7f6778c5cadd346e7/test/functional/rpc_users.py#L159-L162

pinheadmz commented:
concept ACK, this simplifies a confusing behavior with `rpcuser` misconfiguration. Code changes look simple enough, will test locally and run through your (very exhaustive!) behavior comparison

luke-jr commented:
Tested this. Bare `-rpcauth` cannot be used to reverse `-norpcauth`. Since `-rpcauth=somethingvalid` does reverse `-norpcauth` (including reinstating `-rpcauth` params prior to the `-norpcauth`), this seems like it should be supported?

naiyoma commented on test/functional/rpc_users.py:
```diff
@@ -139,15 +139,29 @@ def run_test(self):
         self.nodes[0].assert_start_raises_init_error(expected_msg=init_error, extra_args=['-rpcauth=foo$bar:baz'])
         self.nodes[0].assert_start_raises_init_error(expected_msg=init_error, extra_args=['-rpcauth=foo$bar$baz'])
 
+        # pw = bitcoin
```
nit: Consider adding logs here, makes it easier to understand the test.

tdb3 replied:
Thanks.  Will add after hearing thoughts on treatment of empty rpcauth after norpcauth.

tdb3 replied:
Added log statements. Thanks.

naiyoma commented:
Concept ACK.
 I noticed that [https://github.com/bitcoin/bitcoin/pull/29141/commits/6b79de4f9aed365adda0a99d6522e83ccb6c593f](https://github.com/bitcoin/bitcoin/pull/29141/commits/6b79de4f9aed365adda0a99d6522e83ccb6c593f)  included checks  for `Check -norpcauth can be reversed with -rpcauth` and `Check -norpcauth followed by a specific -rpcauth= restores config file -rpcauth=* values too.` Would it be beneficial to add similar assertions here as well?

tdb3 commented:
> Tested this. Bare `-rpcauth` cannot be used to reverse `-norpcauth`. Since `-rpcauth=somethingvalid` does reverse `-norpcauth` (including reinstating `-rpcauth` params prior to the `-norpcauth`), this seems like it should be supported?

I lean toward empty `rpcauth` always being treated as an error, but I'm open to it being treated differently (at least on CLI) when at least one valid `rpcauth` already exists.
If I'm in the minority, I could experiment with something like the following.

```diff
diff --git a/src/httprpc.cpp b/src/httprpc.cpp
index 9042378a42..6a2245c556 100644
--- a/src/httprpc.cpp
+++ b/src/httprpc.cpp
@@ -317,10 +317,17 @@ static bool InitRPCAuthentication()
 
     if (!gArgs.GetArgs("-rpcauth").empty()) {
         LogInfo("Using rpcauth authentication.\n");
-        for (const std::string& rpcauth : gArgs.GetArgs("-rpcauth")) {
+        size_t num_empty_rpcauths{0};
+        std::vector<std::string> rpcauth_args{gArgs.GetArgs("-rpcauth")};
+        for (const std::string& rpcauth : rpcauth_args) {
             if (rpcauth.empty()) {
-                LogError("Invalid (empty) -rpcauth argument\n");
-                return false;
+                num_empty_rpcauths++;
+                if (num_empty_rpcauths == rpcauth_args.size()) {
+                    LogError("Only invalid (empty) -rpcauth arguments received\n");
+                    return false;
+                } else {
+                    continue;
+                }
             }
             std::vector<std::string> fields{SplitString(rpcauth, ':')};
             const std::vector<std::string> salt_hmac{SplitString(fields.back(), '$')};
```

Thoughts?

ryanofsky commented on src/httprpc.cpp:
```diff
@@ -314,9 +314,14 @@ static bool InitRPCAuthentication()
+    if (!gArgs.GetArgs("-rpcauth").empty()) {
         LogPrintf("Using rpcauth authentication.\n");
         for (const std::string& rpcauth : gArgs.GetArgs("-rpcauth")) {
+            if (rpcauth.empty()) {
+                LogError("Invalid (empty) -rpcauth argument\n");
+                return false;
+            }
```
In commit "fix: increase consistency of rpcauth parsing" (dc38788312e7cdbbc106103e671d548d5c236862)

I think it would be better for code clarity to drop these 4 lines. The only thing they are doing is triggering a special-case "Invalid (empty) -rpcauth argument" error instead of the more generic "Invalid -rpcauth argument." error that would already happen below. The generic error seems sufficient, and If it is not sufficient it would still seem better to handle the two errors together rather than separately.

tdb3 replied:
Thanks.  Will do (i.e. handle the errors together)

tdb3 replied:
Thanks.  Dropped the four lines so the errors are handled together.  Reordered the commits so the tests are first, then the fix.  The tests were adjusted to prevent "test each commit" from failing.

ryanofsky commented:
> Since `-rpcauth=somethingvalid` does reverse `-norpcauth` (including reinstating `-rpcauth` options prior to the `-norpcauth`), this seems like it should be supported?

I think this behavior is surprising and we should not make the code more complicated to support it. It is simpler to just reject empty always, or ignore them always like the previous PR, instead of sometimes accepting or rejecting them based on presence of other values.

For background, the reinstated options are also known as "zombie config values" and are documented in settings code here:

https://github.com/bitcoin/bitcoin/blob/397c6d32c8f8a20a3605ef0d51d159adc21fd125/src/util/settings.cpp#L107-L114

This is also an issue to remove them #17508

ryanofsky reviewed (approved):
Code review ACK 0534fa2c216dbd9e1ae43d30a8a406a0a39c409d but I would really encourage reordering commits in this PR so the tests are added first and the fix is the last commit, so the fix commit updates the tests and you can see what effects it has.

tdb3 committed:
test: add cases for blank rpcauth

tdb3 committed:
test: blank rpcauth CLI interaction

Tests interactions between blank and
non-blank rpcauth args.

tdb3 committed:
test: add norpcauth test

Adds test for disabling rpcauth args.

Co-Authored-By: Luke Dashjr <luke-jr+git@utopios.org>

tdb3 committed:
fix: increase consistency of rpcauth parsing

Previous rpcauth behavior was to sometimes
ignore empty -rpcauth= settings, and other times
treat them as errors.
Empty rpcauth is now consistently treated
as an error and prevents bitcoind from starting.
Updates associated test cases.
Also updates to non-deprecated logging macro.

Co-Authored-By: Luke Dashjr <luke-jr+git@utopios.org>
Co-Authored-By: Ryan Ofsky <ryan@ofsky.org>

tdb3 commented:
Rebased.  Addressed comments from @naiyoma and @ryanofsky.

ryanofsky commented on src/httprpc.cpp:
In commit "fix: increase consistency of rpcauth parsing" (f4b0b69a3e7df58430cb887028b13265834d8c7e)

I think it would be nice if commit message said explicitly that previous behavior was to sometimes ignore empty `-rpcauth=` settings, and other times treat them as errors, and now they are consistently treated as errors.

Could also mention the details, but not important

> Previous behavior was nonsensical:
> 
> - If an empty -rpcauth="" argument was specified as the last command line argument, it would cause all other -rpcauth arguments to be ignored.
> - If an empty -rpcauth="" argument was specified on the command line followed by any nonempty -rpcauth argument, it would cause an error.
> - If an empty "rpcauth=" line was specified after non-empty rpcauth line in the config file it would cause an error.
> - If an empty "rpcauth=" line in a config file was first it would cause other rpcauth entries in the config file to be ignored, unless there were -rpcauth command line arguments and the last one was nonempty, in which case it would cause an error.
> 
> New behavior is simple:
> 
> - If an empty "rpcauth=" config line or -rpcauth="" command line
>   argument is used it will cause an error

tdb3 replied:
Thanks.  Amended the commit message to mention that the previous rpcauth behavior was inconsistent.  Updated the PR description to list detailed behavior.

ryanofsky reviewed (approved):
Code review ACK f4b0b69a3e7df58430cb887028b13265834d8c7e.

This PR is now just a one line fix accompanied by various tests.

ryanofsky reviewed (approved):
Code review ACK 27c976d11a68d66db97d9a7a30c6a6a71c6ab586. Since last review commit message was just tweaked to clarify previous behavior.

I hope this will get reviewed and merged. It is now just a one line fix!

naiyoma commented:
Tested ACK  [https://github.com/bitcoin/bitcoin/commit/27c976d11a68d66db97d9a7a30c6a6a71c6ab586](https://github.com/bitcoin/bitcoin/commit/27c976d11a68d66db97d9a7a30c6a6a71c6ab586)

Thanks for providing the comprehensive list above. I have tested it, and the output for empty `-rpcauth` is as expected: bitcoind fails to start, and Invalid -rpcauth argument is printed.

achow101 commented:
ACK 27c976d11a68d66db97d9a7a30c6a6a71c6ab586
